### PR TITLE
DUPLO-21304 DUPLO-21209 Aurora/v2: replica tries to set value as nil for inplace update and inplace update fixes

### PR DESCRIPTION
**Pull request by Duplo-nikv**
## Overview

Fixed aurora db updation for performance insights state and plan issues
## Summary of changes
Fixed plan issues, 
Fixed replica performance insights issue, 
Updated doc example

This PR does the following:

- Fixed logic at read replica's create and update context for aurora cluster performance insights
- Fixed plan issues
- updated documentation

## Testing performed

- [ ] Using unit tests
- [ ] Manually, on my local system
- [✓ ] Manually, on a remote test system

## Describe any breaking changes

- ...

___

### **PR Type**
enhancement, documentation

___

### **Description**
- Enhanced the RDS instance and read replica schemas by adding `DiffSuppressFunc` to suppress changes when Performance Insights is disabled.
- Improved validation logic for Aurora DB configurations to ensure correct application of Performance Insights settings.
- Updated documentation and examples to reflect changes in Performance Insights configuration, providing clearer guidance on referencing from primary resources.
- Updated the changelog to document the recent enhancements and documentation updates.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>resource_duplo_rds_instance.go</strong><dd><code>Enhance RDS instance schema with diff suppression</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

duplocloud/resource_duplo_rds_instance.go

<li>Added <code>DiffSuppressFunc</code> for <code>kms_key_id</code> and <code>retention_period</code> fields.<br> <li> Improved handling of Performance Insights configuration.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/707/files#diff-a8bf9b56a3e9b39899a5d37d7608110678908754e0848fab2d0320653ea65f71">+34/-6</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>resource_duplo_rds_read_replica.go</strong><dd><code>Improve RDS read replica schema and validation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

duplocloud/resource_duplo_rds_read_replica.go

<li>Added <code>DiffSuppressFunc</code> for <code>kms_key_id</code> and <code>retention_period</code>.<br> <li> Improved validation logic for Aurora DB.<br> <li> Updated <code>CustomizeDiff</code> with <code>validateRDSParameters</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/707/files#diff-1ffdcfb6fc7a0fed517c0f77bbfb406b208cb1084f420e755eb9d029683dfdfe">+35/-31</a>&nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>CHANGELOG.md</strong><dd><code>Update changelog with recent enhancements</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

CHANGELOG.md

<li>Documented enhancements for Performance Insights configuration.<br> <li> Updated documentation for RDS read replicas.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/707/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed">+11/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>rds_read_replica.md</strong><dd><code>Update RDS read replica documentation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docs/resources/rds_read_replica.md

<li>Updated documentation for Performance Insights configuration.<br> <li> Added examples for Aurora cluster level settings.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/707/files#diff-e223dad9b259eae20181173087aafb1d47d2eb221d1a7d6ba1493bd78aafc5af">+26/-7</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>resource.tf</strong><dd><code>Add examples for RDS read replica configurations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

examples/resources/duplocloud_rds_read_replica/resource.tf

<li>Added examples for Performance Insights configuration.<br> <li> Demonstrated Aurora cluster level settings.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/terraform-provider-duplocloud/pull/707/files#diff-3e5f9986f926543549bd2d58be9bb79a2277e56293965198b698e94a55e4d9ff">+26/-7</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Comment by zafarabbas** [comment]
Task linked: [DUPLO-21209 TF:RDS: Update of 'Performance Insight' to false along with 'Retention Period' shows update in place  for plan](https://app.clickup.com/t/8655600/DUPLO-21209)
Task linked: [DUPLO-21304 Aurora/v2: replica tries to set value as nil for inplace update](https://app.clickup.com/t/8655600/DUPLO-21304)

File: duplocloud/resource_duplo_rds_instance.go [hunk 1]
```diff
@@ -1090,13 +1092,39 @@ func performanceInsightsWaitUntilEnabled(ctx context.Context, c *duplosdk.Client
...
+	} else if oldPI.(bool) && newPI.(bool) { //both true check if kms and retention period has change
 		if d.HasChange("performance_insights.0.kms_key_id") || d.HasChange("performance_insights.0.retention_period") {
 			return false
 		}
 		return true
+	} //else if oldPI.(bool) && !newPI.(bool) {
+	//	if d.HasChange("performance_insights.0.kms_key_id") || d.HasChange("performance_insights.0.retention_period") {
+	//		return true
+	//	}
+	//}
```
Comment (Matheusbafutto) [inline duplocloud/resource_duplo_rds_instance.go:1106]: Removed commented out code

File: duplocloud/resource_duplo_rds_read_replica.go [hunk 2]
```diff
@@ -329,26 +327,34 @@ func resourceDuploRdsReadReplicaUpdate(ctx context.Context, d *schema.ResourceDa
...
+	//		return diag.FromErr(insightErr)
+	//
+	//	}
+	//}
+	if !isAuroraDB(d) {
 		insightErr := c.UpdateDBInstancePerformanceInsight(tenantID, obj)
 		if insightErr != nil {
 			return diag.FromErr(insightErr)
 
 		}
```
Comment (Matheusbafutto) [inline duplocloud/resource_duplo_rds_read_replica.go:355]: If it is aurora we should throw an error here, else this TF will wait for instance to become unavailable forever. This section can move out of the conditional block and always execute if not aurora

File: examples/resources/duplocloud_rds_read_replica/resource.tf [hunk 3]
```diff
@@ -65,22 +66,40 @@ resource "duplocloud_rds_instance" "rds" {
...
 resource "duplocloud_rds_read_replica" "replica" {
   tenant_id          = duplocloud_rds_instance.rds.tenant_id
   name               = "read-replica"
   size               = "db.r5.large"
   cluster_identifier = duplocloud_rds_instance.rds.cluster_identifier
+  performance_insights {
+    enabled          = duplocloud_rds_instance.rds.performance_insights.0.enabled
+    retention_period = duplocloud_rds_instance.rds.performance_insights.0.retention_period
+    kms_key_id       = duplocloud_rds_instance.rds.performance_insights.0.kms_key_id
+  }
```
Comment (Matheusbafutto) [inline examples/resources/duplocloud_rds_read_replica/resource.tf:86]: Are we absolutely sure this is the way we want to go about this? If we do it like this we can get rid of this boilerplate https://github.com/duplocloud/terraform-provider-duplocloud/pull/708

Comment (Duplo-nikv) [reply examples/resources/duplocloud_rds_read_replica/resource.tf:86]: I discussed with jeeva on this its the only way

File: duplocloud/resource_duplo_rds_read_replica.go (hunk 2 above)
Comment (Duplo-nikv) [reply duplocloud/resource_duplo_rds_read_replica.go:355]: Ok let me check

Comment (Duplo-nikv) [reply duplocloud/resource_duplo_rds_read_replica.go:355]: I checked the aurora cluster. Before raising PR the block of code skips performance insights updation for aurora. I also checked instance attaching its screenshot 
Screenshot 2024-09-17 at 11 50 22
Screenshot 2024-09-17 at 11 50 33
Screenshot 2024-09-17 at 11 50 49
Screenshot 2024-09-17 at 11 51 24
Screenshot 2024-09-17 at 11 51 34

**Review by Matheusbafutto — APPROVED** [review]

**Review by Matheusbafutto — APPROVED** [review]
looks good to me

**Pull request merged**